src/Network/XMPP/Types.hs

```diff
@@ -378,11 +370,4 @@ data Stanza :: StanzaType -> StanzaPurpose -> * -> * where
         }
         -> Stanza 'IQ p ext
 
-instance Show (Sing 'Incoming) where
-  show _ = "incoming"
-
-instance Show (Sing 'Outgoing) where
-  show _ = "outgoing"
-
-deriving instance (Show ext) => Show (Stanza t 'Incoming ext)
-deriving instance (Show ext) => Show (Stanza t 'Outgoing ext)
+deriving instance (Show (DataByPurpose p ext), Show ext, ShowSing b, Sing p ~ b) => Show (Stanza t p ext)
```

Comment:
Well done! You've resolved this problem quite gracefully 👍🏼

Reply:
Thanks for the encouragement, but doesn't quite work though: https://github.com/riskbook/haskell-xmpp/pull/15/files#diff-e5809c22a3e5b8c46270d7af1985ab57R221
And we're undecidable now: https://github.com/riskbook/haskell-xmpp/pull/15/files#diff-49b715dd8690b25a24036f4923fb54ebR12

I just tried keeping the API somewhat the same, I'm not sure if I succeeded at all.